python/ray/data/block.py

```diff
@@ -458,3 +458,25 @@ def aggregate_combined_blocks(
     def block_type(self) -> BlockType:
         """Return the block type of this block."""
         raise NotImplementedError
+
+
+def _get_block_boundaries(
+    block: Union[np.ndarray, Dict[str, np.ndarray]]
+) -> np.ndarray:
+    """Compute block boundaries based on the key(s), that is, a list of
+    starting indices of each group and an end index of the last group.
+
+    Args:
+        block: numpy arrays of the group key(s). This is generally coming from
+            ``BlockAccessor.to_numpy()``.
+
+    Returns:
+        A list of starting indices of each group. The first entry is 0 and
+        the last entry is ``len(array)``.
+    """
+    if isinstance(block, dict):
+        # For multiple keys, we create a numpy record array
+        dtype = [(k, v.dtype) for k, v in block.items()]
+        block = np.array(list(zip(*block.values())), dtype=dtype)
```

Comment:
There are a few issues with this approach:

1. We're deserializing whole columns into Python objects
2. We need to create structured records np arrays

Let's instead just use `np.lexsort` like we've used to

Reply:
Hmmm...  I agree it's not ideal. Here are some context: `map_groups` sorts and finds group boundaries. Both the original code (`np.searchsorted`) and the new get_boundary code do not work with multiple columns. The conversion to single array was needed in the original multi-column groupby (hence the MultiColumnSortedKey object). The data is already sorted at this point. Maybe we can leave the optimization later? (removal of a while-loop in this pr is already a big win)

Reply:
Oh, good point that it's expected to be already sorted by this point. Let's capture this in a comment that we expect the block to be sorted upon reaching this method.

Then our job is even simpler here we can just do this:

```
np.column_stack(list(d.values()))
```

Reply:
The whole exercise here is to use the fast, vectorized `np.where` to find boundary with one column. Unfortunately it does not allow multiple columns..

Reply:
I'm actually not sure i understand what issue you're referring to.

`column_stack` should produce identical results to what you're trying to do while just avoiding converting objects to Python

Reply:
Ok, i tried to  running it locally and `np.column_stack` does coerce types as transposition creates arrays that have to be homogeneous.

Disregard this comment, let's go ahead with your imp l